### ISSUE TITLE
[explicit-module-boundary-types] allow internal

### ISSUE THREAD

**Issue by mmkal** [typescript-eslint/typescript-eslint#2781]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```typescript
export class MyClass {
  /** @internal */
  getNumber() {
    return 123
  } 
}
```

```JSON
{
  "rules": {
    "@typescript-eslint/<rule>": ["@typescript-eslint/explicit-module-boundary-types"]
  }
}
```

tsconfig (only [`stripInternal`](https://www.typescriptlang.org/tsconfig#stripInternal) is important):
```json
{
	"include": [
		"src"
	],
	"compilerOptions": {
		"stripInternal": true,

		"outDir": "lib",
		"target": "es2018",
		"module": "commonjs",
		"moduleResolution": "node",
		"resolveJsonModule": true,
		"declaration": true,
		"sourceMap": true,
		"pretty": true,
		"newLine": "lf",
		"strictNullChecks": true,
		"noImplicitAny": true,
		"noImplicitReturns": true,
		"noUnusedLocals": false,
		"noUnusedParameters": false,
		"noFallthroughCasesInSwitch": true,
		"noEmitOnError": true,
		"skipLibCheck": true
	}
}
```

**Expected Result**

An option like `allowInternal` that gets rid of the error.

**Actual Result**

Error

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.8.1` |
| `@typescript-eslint/parser`        | `4.8.1` |
| `TypeScript`                       | `4.0.5` |
| `ESLint`                           | `7.13.0` |
| `node`                             | `v12.18.0` |

**Comment by bradzacher** [comment]
I disagree that `@internal` should be ignored.
`@internal` means that one specific type is not emitted into the type declaration.

It does not mean that it does not make up the signature of the file.

Whilst yes - if you compile the file such that its only signature is its `.d.ts` file, then the `@internal` type not there.

However internal to the bundle - within the package, the `@internal` type is still part of the publicly available signature.


```ts
// ./file1.ts

export function foo() {
  return 'foo';
}

// @internal
export function bar() {
  return 'bar';
}

/////

// ./file2.ts

import {foo, bar} from './file1'; // 

foo(); // both internal and external
bar(); // only internal - but still a visible type to this module
```

**Comment by mmkal** [comment]
I agree, it shouldn't be ignored in general. But it might be safe, and even helpful, in certain projects. For example in large, non-public monorepos, often the unit of abstraction is an npm package rather than a `.ts` module file. The rough development pattern I've seen there is "explicit types for anything exported, inference is allowed for anything else". And it can be pretty useful to take advantage of the compiler's type inference for complex types. But you are right it'd be a bad idea to turn on for everyone, so I've edited the OP to make it a request for the `allowInternal` option.

Re `file2.ts` in your example, I'm not sure I follow. It looks like `foo()` and `bar()` are just side-effects, so wouldn't have types anyway? And if the return values were exported, they'd still be subject to the rule:

```typescript
// ./file2.ts

import { foo, bar } from './file1'

export const fooResult = foo(); // explicit-module-boundary-types error
export const barResult = bar(); // still an explicit-module-boundary-types error under this proposal, since even though `bar` is internal, `barResult` is missing a type
```

**Comment by bradzacher** [comment]
> Re file2.ts in your example, I'm not sure I follow

For simplicity's sake, let me define two terms:
- "package" - by this I mean all of the files included and bundled with `file1.ts`
- "module" - by this I mean a single file, eg `file1.ts`

The idea I was trying to convey with my example was not to do with the linting, but rather with the concept I was conveying.
An `@internal` type is part of the contract provided by the module to consumers within the package.
Whilst yes, a consumer external to the package would not see an `@internal` type - **this is not the level that the rule operates at** - the rule isn't `explicit-package-boundary-types`, it's `explicit-module-boundary-types`. The rule explicitly operates at the module level.

To add a counter example - imagine the following package setup:

```ts
// ./package/file1.ts

export function getFoo(): string {
  return 'foo';
}

// ./package/index.ts

import { getFoo } from './file1';

export default getFoo();

// ./other-package/file2.ts

import package from 'package';

console.log(package); // -> 'foo'
```

In this example - `file1.ts` is entirely internal to the package. Without the need for any `@internal` annotations, its types are hidden and are not exposed by the package.

What should the lint rule do here? Should it analyse the package tree to decide that `file1.ts` is entirely internal and not require boundary types?

The answer is no - because the rule does not operate at the package level. It operates specifically at the module level.

---

> here's a hypothetical example of where an explicit type adds overhead without much value

I've always been a strong proponent of strict contracts - triply so at module boundaries.
Providing clear contracts to consumers helps set expectations and prevent strange bugs from leaking around the place.

Many people do feel like certain types are hard to manually define, which is why many people do not like this rule or the `explicit-function-return-types` rule.

My answer to those is the same: Do not blindly follow the linter and make your life harder.
If you think that you have a legitimate usecase where the type is too complicated to manually define, then don't. Suppress the linter with a clear comment and move on.

```ts
/** @internal */
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types -- getOptions has a very big, complex type signature, but is only used internally so we're happy to have it be inferred
export const getOptions => () => ({
  fooEnabled: process.env.FOO === 'true',
  barEnabled: process.env.BAR === 'true',
  bazEnabled: process.env.BAZ === 'true',
  ...many more
}) 
```

Too many people see `eslint-disable` comments as a cardinal sin.
But it's 100% okay to disable the linter if you have a good reason and a clear comment.

**Comment by mmkal** [comment]
> the rule isn't explicit-package-boundary-types, it's explicit-module-boundary-types

Very fair - maybe what I'm really requesting here is a _new_ rule, `explicit-package-boundary-types`? Where to keep things simple (avoiding trying to parse `package.json` for all the files that are part of the API surface), the only real difference is the leniency over `internal`-marked properties, so implementation-wise could be a bare-bones extension of this one.

For your `export default` example - I'd expect that to be a violation. I'm not a big fan of `export default something` where `something` is imported. It's not easy to tell what that `something` is.

I also agree about eslint-disable comments. But I usually see them as a stopgap. If they start popping up frequently, it's usually a sign that the org is converging on a convention that isn't covered by the existing set of rules. That's what happened in this case. I've noticed myself and others disabling explicit-module-boundary-types somewhat frequently. I had a look at the cases where they were _bad_ suppressions (laziness) vs _good_ suppressions (increasing maintainability). I found that the _good_ ones were internal (whether marked that way or not). So my first thought was to see if the rule could be reconfigured to allow internal. When it couldn't, I came here requesting an enhancement!

Would you be open to a separate rule that was otherwise the same, but allowed internal properties?

**Comment by bradzacher** [comment]
> Would you be open to a separate rule for packages, that was otherwise the same, but allowed internal properties

I don't think that it should live here in this plugin. I don't think there's enough drive for one - the majority of people work and think about this stuff at the module level..

As this project has grown, I've tended towards pushing back against changes that I don't think will be widely adopted. Each new rule / option has a maintenance cost, and as the maintainers are volunteers - I'm conscious of the cost:value ratio.

You're welcome to fork the rule and update it for your purposes as allowed by the [MIT Licence](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/LICENSE) (whether or not you open-source the rule or just use it internal to your project is up to you).

**Comment by mmkal** [comment]
Fair enough!
